src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,546 @@
+    ProposalInfo storage $ = _info[hash];
+
+    if ($._votePow.vAgainst >= _calcMinVoteAgainst($._totalVotePow)) {
+      $._state = State.Cancelled;
+      emit ProposalCancelled(msg.sender, hash);
+      return false;
+    }
+
+    if ($._votePow.vFor < _calcMinVoteFor($._totalVotePow)) return false;
+
+    // Removal can be skipped if it's already removed when proposal is passed by voting but not automatically executed.
+    _pendingProposals.remove(hash);
+
+    Proposal storage $p = $._proposal;
+
+    // Skip executing if `executor` is not the contract itself or `msg.sender`.
+    if (!($p.executor == address(this) || $p.executor == msg.sender)) {
+      $._state = State.Passed;
+      emit ProposalPassed(msg.sender, hash);
+      return false;
+    }
+
+    uint256 length = $p.recipients.length;
+
+    for (uint256 i; i < length; ++i) {
+      (bool success, bytes memory ret) = $p.recipients[i].call{ value: $p.amounts[i] }($p.callDatas[i]);
+      success.handleRevert(msg.sig, ret);
+    }
+
+    $._state = State.Executed;
+
+    emit ProposalExecuted(msg.sender, hash);
+
+    return true;
+  }
+
+  /**
+   * @dev Set the threshold for passing proposals.
+   *
+   * - Increment the global nonce.
+   *
+   * Emits a {GlobalNonceUpdated} event.
+   * Emits a {ThresholdUpdated} event.
+   */
+  function _setThreshold(uint8 num, uint8 denom) internal {
+    _incrementNonce();
+
+    require(!(denom == 0 || num == 0 || num >= denom), ErrInvalidThreshold(num, denom));
+
+    _threshold = Threshold(num, denom);
+
+    emit ThresholdUpdated(msg.sender, num, denom);
+  }
+
+  /**
+   * @dev Increment the global nonce.
+   *
+   * Emits a {GlobalNonceUpdated} event.
+   */
+  function _incrementNonce() internal {
+    ++_globalNonce;
+    emit GlobalNonceUpdated(msg.sender, getGlobalNonce());
+  }
+
+  /**
+   * @dev Try to remove the pending proposals.
+   */
+  function _syncPendingProposals() internal {
+    uint256 length = _pendingProposals.length();
+    bytes32[] memory pendingProposals = _pendingProposals.values();
+
+    for (uint256 i; i < length; ++i) {
+      if (getState(pendingProposals[i]) != State.Active) {
+        _pendingProposals.remove(pendingProposals[i]);
+      }
+    }
+  }
+
+  /**
+   * @dev Calculate the minimum vote power required for a proposal to pass.
+   */
+  function _calcMinVoteFor(
+    uint256 totalVotePower
+  ) internal view returns (uint96 minVoteFor) {
+    minVoteFor = ((totalVotePower * _threshold.num) / _threshold.denom).toUint96();
+    require(minVoteFor != 0, ErrZeroMinVotePower());
+  }
+
+  /**
+   * @dev Calculate the minimum vote power required for a proposal to cancel.
+   */
+  function _calcMinVoteAgainst(
+    uint256 totalVotePower
+  ) internal view returns (uint96 minVoteAgainst) {
+    uint256 diff = _threshold.denom - _threshold.num;
+    minVoteAgainst = ((totalVotePower * diff) / _threshold.denom).toUint96();
+    require(minVoteAgainst != 0, ErrZeroMinVotePower());
```

Comment:
This function should be calculated by taking `totalVotePower - _calcMinVoteFor`. Because with your implementation, may `_calcMinVoteFor + _calcMinVoteAgainst < totalVotePower`, which leads to a situation where the proposal is canceled once the votes just meet `_calcMinVoteAgainst`, even though further voting could potentially exceed `_calcMinVoteFor`.

Reply:
Agree. To be more specific, let’s say the total vote power is `31` and the threshold is `0.5` (num = 1, denom = 2). Both `_calcMinVoteAgainst` and `_calcMinVoteFor` would return 15. Even if the current power of `voteFor` is 16, the proposal would still be canceled

Reply:
acked. However, if it is actually come in this case on production, I still think it is okay if it is cancelled.